[Refactor] pickeat 플로우에서 인메모리 캐싱 적용

jinu0328 opened the pull request:
## Issue Number
#407 

기존 객체참조 부분을 모두 간접참조로 바꾸다보니 변경이 전파된 부분이 많습니다 ㅠㅠ
커밋별로 확인해주시는 편이 파악하시기 수월할 것 같아요.
커밋별로 한 작업 간단히 정리해봤습니당
flyway 관련해서는 로컬에서 테스트 완료하긴했는데 검수 한번씩 꼭 부탁드려요!!
질문 지적 환영합니다~~

### 기반 세팅

2935a851 Spring Cache + Caffeine 의존성 추가
→ 이후 전역 캐싱 도입을 위한 기반 세팅입니다.

### 전역 캐시 구성

a79ad1c0 Caffeine 기반 글로벌 CacheManager 구성
→ TTL·사이즈·캐시 이름을 한 곳에서 관리하도록 중앙화했습니다.

### 도메인 구조 정리

c52ad635 Restaurant 계열 엔티티가 Pickeat을 Long ID로 간접 참조하도록 리팩토링
→ 양방향 연관 제거로 JPA 로딩/삭제 부담을 줄였습니다.

4df9574c Participant 엔티티도 Pickeat을 Long ID로 간접 참조하도록 변경
→ 순환 참조 위험을 줄이고 구조를 단순화했습니다.

830b09b3 PickeatResult에서 Pickeat/Restaurant 직접 필드를 Long ID로 교체
→ 도메인 전반을 간접 참조 방식으로 통일했습니다.

69744680 Pickeat의 참가자 수 필드 및 로직 제거
→ 스키마와 도메인 상태를 일관되게 유지했습니다.

508c2ed2 PickeatResult의 동점 여부 필드 제거 및 관련 로직 정리
→ 사용되지 않는 필드를 정리해 도메인 단순화.

0b22cee7 Restaurants 도메인을 PickeatResultGenerator로 대체
→ 레스토랑 선택 로직을 별도 서비스로 분리하여 테스트 용이성 및 응집도 향상.

### 캐시 도입 및 최적화

bc6a0062 RestaurantRepositoryImpl에 Pickeat별 식당 조회 캐시 적용
→ 기존 도메인 기반 조회 대신 캐시 결과를 사용하도록 개선.

20e32f89 RestaurantLikeRepositoryImpl 캐싱 도입
→ 좋아요 수·참여자별 좋아요 목록 캐싱, save/delete 시 캐시 무효화 처리 추가.

13ca77e1 PickeatRepositoryImpl 캐싱 적용
→ 코드 기반 Pickeat 조회에 캐시 추가 및 비활성화 시 캐시 무효화.

ab464e9c ParticipantRepositoryImpl에 Spring Cache 적용
→ Pickeat별 참가자 목록 캐싱 및 저장/조회 로직 리팩토링.

### 이벤트 기반 캐시 무효화

eeb93d6b Pickeat 비활성화 이벤트 신설
→ Pickeat 비활성화 시 식당·Pickeat 캐시를 트랜잭션 이벤트로 통합 처리.

a047a2ac Pickeat 비활성화 시 참가자 캐시도 함께 비우도록 이벤트 핸들러 보강
→ 기존 Pickeat/식당 캐시만 비워지던 문제를 해결.

9e44b1c8 참가자 완료 상태 변경 이벤트 및 캐시 무효화 핸들러 추가
→ 참가자 상태 업데이트 시 캐시가 즉시 갱신되도록 개선.

### Flyway 및 DB 정합성

fd54f64e PickeatResult·Restaurant 간 직접 연관을 Long ID 기반으로 변경
→ Flyway V10~V14 스키마 정리 및 인덱스 재구성 포함.


## Check List
- [x] 빌드, 테스트가 전부 통과되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?


## (Optional) Additional Description


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

# 릴리스 노트

* **새로운 기능**
  * 캐싱 레이어 추가로 API 응답 성능 개선

* **버그 수정**
  * 인증 토큰 생성 로직 안정성 개선
  * 응답 로그에서 불필요한 바디 데이터 제거

* **개선 사항**
  * 데이터 접근 패턴 최적화
  * 이벤트 기반 캐시 무효화 메커니즘 구현
  * 내부 데이터 모델 정규화
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jinu0328 committed:
feat: Spring Cache 및 Caffeine 의존성 추가

- 캐시 활용을 위한 spring-boot-starter-cache 및 caffeine 의존성 추가

jinu0328 committed:
refactor: ResponseLog에서 응답 바디 관련 코드 제거

- 로깅 간결화를 위해 body 필드 제거

jinu0328 committed:
refactor: logback-local.xml에 출력 포맷 변경

- LoggingEventCompositeJsonEncoder를 활용하여 JSON 형식의 로그 출력 적용
- 로깅 타임스탬프, 레벨, 쓰레드명, 메시지 등의 정보를 구조화하여 기록

jinu0328 committed:
refactor: 참가자 토큰에 pickeatCode 클레임 추가 및 기존 토큰 생성 로직 리팩토링

- ParticipantTokenProvider와 ProviderTokenProvider 생성 및 로직 분리
- JwtProvider 리팩토링으로 추가 정보를 포함한 토큰 생성 기능 도입
- 관련 컨트롤러 및 서비스 로직 수정

jinu0328 committed:
refactor: 인증된 사용자 정보 info -> principal로 리네이밍

jinu0328 committed:
chore: 투두 작성

jinu0328 committed:
refactor: Pickeat 엔티티에서 참가자 수 관련 필드 및 로직 제거

- Pickeat 엔티티의 participantCount 필드 삭제

jinu0328 committed:
refactor: PickeatResult 관련 동점 여부 로직 제거 및 코드 정리

- hasEqualLike 관련 로직 삭제
- RestaurantResultResponse -> PickeatResultResponse로 클래스명 변경
- PickeatResult에 동점 여부 필드 제거
- SQL 및 테스트 코드 수정
- 불필요한 코드를 간소화하고 가독성을 개선

jinu0328 committed:
refactor: PickeatResult에서 Pickeat 및 Restaurant 엔티티 직접 참조를 Long ID 기반 간접참조로 대체

- Pickeat 및 Restaurant를 ID로만 저장하도록 수정
- Repository 및 Service 로직 변경
- 테스트 코드 수정 및 불필요한 OneToOne 매핑 제거

jinu0328 committed:
refactor: Participant에서 Pickeat 엔티티 직접 참조를 Long ID 기반으로 변경

- Pickeat을 ID로만 저장하도록 수정
- Repository 및 Service 로직 개선
- 테스트 코드 수정 및 불필요한 ManyToOne 매핑 제거
- 코드 가독성 향상 및 성능 최적화

jinu0328 committed:
refactor: Restaurant 및 관련 엔티티에서 Pickeat 직접 참조를 Long ID 기반으로 변경

- Pickeat을 ID로만 저장하도록 수정
- Restaurant의 likeCount 및 관련 로직 제거
- RestaurantService 및 Repository 로직 변경 및 개선
- Tests 내부 로직 및 픽스처 생성 방식 수정
- 기존 Restaurants 도메인 클래스 제거
- 코드 가독성 향상 및 무결성 검증 로직 단순화

jinu0328 committed:
refactor: Restaurants -> PickeatResultGenerator로 변경

- 단위 테스트 불가한 도메인 제거 후 애플리케이션 레이어로 이동

jinu0328 committed:
feat: Caffeine 기반 글로벌 캐시 매니저 설정 추가

- CacheManagerConfig 클래스 생성 및 글로벌 캐시 매니저 설정
- CacheType enum 정의 및 캐시 타입별 속성 관리
- RESTAURANT 캐시 타입 추가 (TTL, 최대 크기, 초기 용량 설정 포함)

jinu0328 committed:
refactor: 픽잇의 Restaurant 조회시 캐싱 적용

- RestaurantRepository를 JpaRepository와 JdbcTemplate 기반으로 분리해 RestaurantJpaRepository와 RestaurantJdbcRepository로 구현
- Pickeat 관련 캐싱 설정 도입 및 삭제 시 캐시 무효화 로직 적용

jinu0328 committed:
feat: Pickeat 비활성화 이벤트 처리 로직 추가 (캐시 무효화)

- Pickeat 비활성화 시 발생하는 PickeatDeactivatedEvent 생성
- 해당 이벤트를 처리하여 Restaurant 캐시를 무효화하는 PickeatDeactivatedEventHandler 추가

jinu0328 committed:
feat: RestaurantLike 관련 캐싱 적용

- RestaurantService에서 사용자 좋아요 데이터 조회 최적화
- RestaurantLike 케싱 관련 테스트 코드 추가

jinu0328 committed:
feat: Pickeat 캐싱 및 관련 리팩토링 적용

- PickeatRepository 인터페이스 분리 및 PickeatRepositoryImpl 클래스 생성
- Pickeat 검색 결과 캐싱 로직 추가 (Caffeine 활용)
- Pickeat 캐시 사용 테스트 코드 추가
- PickeatJpaRepository로 리팩토링 및 관련 로직 최적화
- Pickeat 비활성화 시 캐시 무효화 처리 로직 강화 (PickeatDeactivatedEventHandler 수정)

jinu0328 committed:
feat: 참가자 캐싱 및 관련 리팩토링 적용

- ParticipantRepository 인터페이스 분리 및 ParticipantRepositoryImpl 클래스 생성
- 참가자 검색 결과 캐싱 로직 추가 (Caffeine 활용)
- 참가자 캐시 사용 테스트 코드 추가
- ParticipantJpaRepository로 리팩토링 및 관련 코드 최적화

jinu0328 committed:
feat: 참가자 상태 업데이트 이벤트 처리 로직 추가

- ParticipantStateUpdatedEvent 도메인 이벤트 클래스 생성
- ParticipantStateUpdatedEventHandler 구현 및 트랜잭션 후 이벤트 처리 로직 추가
- 참가자 캐시 무효화 로직 적용 (ParticipantRepositoryImpl 활용)

jinu0328 committed:
feat: Pickeat 비활성화 시 참가자 캐시 무효화 로직 추가

- PickeatDeactivatedEventHandler에 참가자 캐시 무효화 처리 로직 적용
- PickeatRepositoryImpl의 캐싱 키 표현식 수정 (value.toString() 사용)

jinu0328 committed:
refactor: PickeatResult 및 Restaurant 간 간접 참조를 위한 스키마 개편

- PickeatResult에서 has_equal_like 컬럼 제거
- PickeatResult 테이블의 FK 제약 조건 삭제 및 pickeat_id, restaurant_id 필드 NOT NULL로 수정
- Restaurant에서 like_count, type 컬럼 제거 및 기존 외래키 제약 조건 삭제
- Participant 및 Restaurant 테이블에 대한 새로운 인덱스 생성
- 간접 참조 처리 기반 데이터 무결성 유지 및 불필요한 컬럼 정리

jinu0328 committed:
fix: Pickeat 비활성화 로직에서 Detached된 엔티티 명시적으로 save 호출

-  캐시가 JPA 엔티티를 그대로 저장하면서 수정 트랜잭션에서 Detached 엔티티가 사용되
    고, 이 때문에 비활성화/상태 변경이 DB에 반영되지 않는 문제 명시적 save 호출로 해결

jinu0328 committed:
refactor: JWT 토큰 생성 로직 개선 및 사용자 정의 클레임 추가 처리

- 사용자 정의 클레임(extraClaims) 추가 시 예약 클레임(sub, iat, exp) 덮어쓰기 방지 로직 적용
- JwtBuilder의 생성 순서를 명확히 구분 (사용자 정의 클레임 → JWT 예약 클레임 → 최종 토큰 생성)
- 토큰 생성 관련 코드의 가독성과 안정성 향상

jinu0328 committed:
refactor: 캐시 이름 상수화 및 CacheNames 인터페이스 도입

- 캐시 이름을 상수로 관리하기 위한 CacheNames 인터페이스 추가
- 기존 캐시 이름을 CacheNames 상수로 대체하여 코드 일관성 및 유지보수성 향상
- 관련된 Repository 구현체의 캐싱 키 설정 코드 리팩토링 (Restaurant, Participant, Pickeat 등)

jinu0328 committed:
refactor: Restaurant 제외 로직에서 Boolean 비교 방식 개선

- Boolean.FALSE.equals 사용으로 NPE 방지 및 코드 가독성 향상

jinu0328 commented on backend/src/main/java/com/pickeat/backend/pickeat/application/PickeatScheduler.java:
```diff
@@ -25,11 +26,11 @@ public class PickeatScheduler {
+    private final ParticipantJpaRepository participantJpaRepository;
+    private final RestaurantLikeJpaRepository restaurantLikeRepository;
+    private final RestaurantJpaRepository restaurantJpaRepository;
 
```
@coderabbitai
스케줄러는 캐시 생성 트리거 자체를 호출하지 않기 때문에,
애초에 해당 시점에 캐시에 데이터가 존재하지 않는 것 같아.
따라서 캐시 무효화 이벤트를 발행할 필요도 없고,정합성 측면에서 문제 없는 구조같은데 어떻게 생각해

jinu0328 committed:
fix: 테스트 코드 내 캐시 이름 상수화 적용

Commit:
refactor: 스케쥴링 대량 삭제 로직에서 bulk delete 사용

wodnd0131 commented on backend/src/main/java/com/pickeat/backend/pickeat/application/PickeatScheduler.java:
```diff
@@ -49,17 +50,20 @@ public void cleanupOldPickeats() {
         pickeatResultRepository.deleteAll(resultsToDelete);
-        participantRepository.deleteAll(participantsToDelete);
+        participantJpaRepository.deleteAll(participantsToDelete);
     }
```
현재 `@SoftDelete`를 사용한 소프트 딜리트 설정으로 인하여, JPA를 사용한 엔티티 삭제를 인터셉트 되어집니다.
이는 Delete를 Update로 바꾸는 작업을 의미하며, 위와 같이 `deleteByPickeatIdIn`를 사용해도 벌크연산이 아닌 개별 update 연산을 변경됩니다.

때문에 벌크 연산을 위해서 별도의 native query를 통한 벌크 쿼리와 관련 테스트를 추가했습니다.
(JPQL의 경우, deleted 필드가 명시되지 않아 인식되지 않습니다.)

Commit:
refactor: PickeatDeactivatedEvent의 필드를 id와 code로 세분화

wodnd0131 commented on backend/src/main/java/com/pickeat/backend/pickeat/infrastructure/ParticipantRepositoryImpl.java:
```diff
@@ -0,0 +1,47 @@
+    @Cacheable(value = PARTICIPANT_CACHE_NAME, key = "#pickeatId")
+    public List<Participant> findByPickeatId(Long pickeatId) {
+        return jpaRepository.findByPickeatId(pickeatId);
+    }
```
`@CacheConfig` 처리는 채택, null 체크는 dto 단에서 진행하는 컨밴션을 지킨다.

Commit:
refactor: 클래스별 캐시 설정을 CacheConfig로 통합

- createPickeatResult 명세 최신화

Commit:
fix: 비활성화 API 경로에 오타 수정 및 호환성 문제 예방

Commit:
fix: PickeatController에서 참가자 ID를 null 체크하여 재참여 가능성 조회 로직 수정

Commit:
refactor: 메서드 이름 변경 및 불필요한 메서드 제거

- RestaurantLikeJpaRepository에서 countAllByRestaurantId를 countByRestaurantId로 변경
- RestaurantJpaRepository에서 findIdsByPickeatIdIn 메서드 제거 및 findByPickeatIdIn으로 변경
- 관련 테스트에서 메서드 이름 변경에 따른 수정

wodnd0131 commented on backend/src/main/java/com/pickeat/backend/pickeat/infrastructure/PickeatRepositoryImpl.java:
```diff
@@ -0,0 +1,65 @@
+    @CacheEvict(value = PICKEAT_CACHE_NAME, key = "#pickeat.code.value.toString()")
+    public void evictPickeatCache(Pickeat pickeat) {
+        log.info("픽잇 캐시 무효화 | pickeatCode: {}", pickeat.getCode().getValue().toString());
+    }
```
없는 걸로 훈수두지마

Commit:
refactor: Pickeat 접근 권한 검증 및 Restaurant 저장 로직 수정

supernovaMK commented on backend/src/main/java/com/pickeat/backend/tobe/restaurant/ui/RestaurantController.java:
```diff
@@ -65,29 +66,29 @@ public ResponseEntity<Void> createRestaurantsByTemplate(
     }
 
     @Override
     @PatchMapping("/restaurants/{restaurantId}/like")
```
[P2]
like,cacelLikeRestaurants 가가 Patch인게 이제 어색해진 것 같다는 생각이 듭니다.
어떻게 생각하시나요?

supernovaMK commented on backend/src/main/java/com/pickeat/backend/restaurant/infrastructure/RestaurantRepositoryImpl.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public Optional<Restaurant> findById(Long restaurantId) {
+        return jpaRepository.findById(restaurantId);
+    }
```
단일 검색에서는 cache를 사용하고 있지 않네요! 이유가 궁금합니다!

cahche되어있는 List<Restaurant> 에서 검색하는 것보다 이게 낫다고 판단하신 걸까요?

supernovaMK reviewed (approved):
에드~ 수고 많으셨습니다!

요구사항이 많았는데 모두 잘 해결해주셨네요 . 몇가지 코멘트 달아두었습니다!

jinu0328 committed:
refactor: RestaurantLikeCount 도입으로 좋아요 수 관리 로직 개선

- 좋아요 수를 관리하기 위해 RestaurantLikeCount 클래스 추가
- 기존 int 기반 카운트를 AtomicLong 기반의 RestaurantLikeCount로 대체
- 좋아요 및 좋아요 취소 시 카운트 증가/감소 처리 로직 추가
- 캐싱 로직 수정으로 데이터 일관성 강화 및 코드 가독성 향상
- 관련 테스트 코드 및 Repository 구현체 수정

jinu0328 committed:
refactor: 좋아요 수 관리 로직 간소화 및 ParticipantLikes 기반 리팩토링

- RestaurantLikeCount 클래스 및 관련 로직 제거
- 좋아요 수 관리를 ParticipantLikesRepository로 통합
- ParticipantLikes를 활용한 좋아요 카운트 및 상태 관리로 코드 간소화
- 캐시 및 테스트 코드 수정으로 불필요한 캐싱 제거 및 유지보수성 개선
- Pickeat 생성 시 캐시 초기화 로직 적용
- 관련 인터페이스 및 Repository 리팩토링

jinu0328 committed:
Merge branch 'back/dev' into 407-feat-pickeat-플로우에서-인메모리-캐싱-적용

Commit:
fix: 수정된 변수 이름으로 일관성 있는 메서드 정의
